velox/expression/Expr.h

```diff
@@ -331,6 +331,10 @@ class Expr {
   // parent Expr.
   std::vector<FieldReference * FOLLY_NONNULL> distinctFields_;
 
+  // Fields referenced by multiple inputs, which is subset of distinctFields_.
+  // Used to determine pre-loading of lazy vectors at current expr.
+  std::unordered_set<FieldReference * FOLLY_NONNULL> multiRefFields_;
```

Comment:
`distinctFields_` and `multiRefFields_` could be merged into one field `F14FastMap<FieldReference * FOLLY_NONNULL, int> fieldRefCount_`.

Reply:
If distinctFields_ is same with parent expr's distinctFields_, then distinctFields_ will be clear.
While multiRefFields_ needn't to compare with its parent expr's multiRefFields_.

Reply:
OK I see, `distinctFields()` are used in other places so better keep it this way.
Another question though, since `multiRefFields_` is only used in root, is it possible to make it only non-empty in root (and document it)?

Reply:
I see the point to keep it in class.
For eliminate the redundancy, can you do it like `distinctFields_`, where we drop the values in the child when we compute the value in parent?

Reply:
To eliminate the redundancy, it needs a traversal of entire sub-expression trees at each parent expr node.  
It's a tradeoff between memory space and time.

Reply:
If you leave `multiRefFields_` in child nodes, they will be checked for each batch during eval.

To remove the redundancy, is it sufficient to have a single traversal of the tree while keeping a running union of `multiRefFields_` on all ancestors of current node?

Reply:
@Yuhta I got the algorithm you described to remove the redundancy. But, it needs an extra traversal after `Expr:: computeMetadata()`.  
An expr node need to visit all sub-expression trees, then find out the `multiRefFields_ `.  

`multiRefFields_ ` actually will be checked for each batch during eval. And will quickly bypass loading vector if it was already loaded. Compared with computation of scalar functions for thousands rows in the batch,  the cost can be ignored.  
Similar redundancy also exists for `distinctFields_ `. e.g. expression `strpos(upper (a), ‘FOO’) > 0 OR strpos(upper(b), BAR) > 0`. Both `OR` node and `>` node will keep field a. And they will also be checked for each batch during eval.  

So, I prefer to keep the redundancy, and make the logic of `Expr::computeMetadata ()` simple.

Reply:
OK we can live with that for now.